registry/storage/manifeststore.go

```diff
@@ -102,6 +103,48 @@ func (ms *manifestStore) verifyManifest(mnfst *manifest.SignedManifest) error {
 		}
 	}
 
+	if len(mnfst.FSLayers) == 0 || len(mnfst.History) == 0 {
+		errs = append(errs, distribution.ErrManifestValidation{
+			Reason: "no layers present"})
+	}
+
+	if len(mnfst.FSLayers) != len(mnfst.History) {
+		errs = append(errs, distribution.ErrManifestValidation{
+			Reason: "mismatched layers and history"})
+	}
+
+	// image provides a local type for validating the image relationship.
+	type image struct {
+		ID     string `json:"id"`
+		Parent string `json:"parent"`
+	}
+
+	// Process the history portion to ensure that the parent links are
+	// correctly represented. We serialize the image json, then walk the
+	// entries, checking the parent link.
+	var images []image
+	for _, entry := range mnfst.History {
+		var im image
+		if err := json.Unmarshal([]byte(entry.V1Compatibility), &im); err != nil {
+			errs = append(errs, err)
+		}
+
+		images = append(images, im)
+	}
+
+	// go back through each image, checking the parent link and rank
+	var parentID string
+	for i := len(images) - 1; i >= 0; i-- {
+		// ensure that the parent id matches but only if there is a parent.
+		// There are cases where successive layers don't fill in the parents.
+		if images[i].Parent != parentID {
```

Comment:
I'm seeing this check fail with a `hello-world` image. Its manifest has a duplicated layer:

    {
       "name": "hello-world",
       "tag": "latest",
       "architecture": "amd64",
       "fsLayers": [
          {
             "blobSum": "sha256:a3ed95caeb02ffe68cdd9fd84406680ae93d633cb16422d00e8a7c22955b46d4"
          },
          {
             "blobSum": "sha256:a3ed95caeb02ffe68cdd9fd84406680ae93d633cb16422d00e8a7c22955b46d4"
          },
          {
             "blobSum": "sha256:9a57efb1f998eefdfdea1128834ffca5b3bc189a780c172d98069c34e487af7d"
          }
       ],
       "history": [
          {
             "v1Compatibility": "{\"id\":\"91c95931e552b11604fea91c2f537284149ec32fff0f700a4769cfd31d7696ae\",\"parent\":\"a8219747be10611d65b7c693f48e7222c0bf54b5df8467d3f99003611afa1fd8\",\"created\":\"2015-04-17T22:01:38.74474669Z\",\"container\":\"724582a033a54c0864aab1e37d7bfaea9a3228bc63282df0dd07dd5768da40be\",\"container_config\":{\"Hostname\":\"d7c065bb52d8\",\"Domainname\":\"\",\"User\":\"\",\"AttachStdin\":false,\"AttachStdout\":false,\"AttachStderr\":false,\"PortSpecs\":null,\"ExposedPorts\":null,\"Tty\":false,\"OpenStdin\":false,\"StdinOnce\":false,\"Env\":null,\"Cmd\":[\"/bin/sh\",\"-c\",\"#(nop) CMD [\\\"/hello\\\"]\"],\"Image\":\"a8219747be10611d65b7c693f48e7222c0bf54b5df8467d3f99003611afa1fd8\",\"Volumes\":null,\"VolumeDriver\":\"\",\"WorkingDir\":\"\",\"Entrypoint\":null,\"NetworkDisabled\":false,\"MacAddress\":\"\",\"OnBuild\":null,\"Labels\":{}},\"docker_version\":\"1.6.0\",\"config\":{\"Hostname\":\"d7c065bb52d8\",\"Domainname\":\"\",\"User\":\"\",\"AttachStdin\":false,\"AttachStdout\":false,\"AttachStderr\":false,\"PortSpecs\":null,\"ExposedPorts\":null,\"Tty\":false,\"OpenStdin\":false,\"StdinOnce\":false,\"Env\":null,\"Cmd\":[\"/hello\"],\"Image\":\"a8219747be10611d65b7c693f48e7222c0bf54b5df8467d3f99003611afa1fd8\",\"Volumes\":null,\"VolumeDriver\":\"\",\"WorkingDir\":\"\",\"Entrypoint\":null,\"NetworkDisabled\":false,\"MacAddress\":\"\",\"OnBuild\":null,\"Labels\":{}},\"architecture\":\"amd64\",\"os\":\"linux\",\"Size\":0}\n"
          },
          {
             "v1Compatibility": "{\"id\":\"91c95931e552b11604fea91c2f537284149ec32fff0f700a4769cfd31d7696ae\",\"parent\":\"a8219747be10611d65b7c693f48e7222c0bf54b5df8467d3f99003611afa1fd8\",\"created\":\"2015-04-17T22:01:38.74474669Z\",\"container\":\"724582a033a54c0864aab1e37d7bfaea9a3228bc63282df0dd07dd5768da40be\",\"container_config\":{\"Hostname\":\"d7c065bb52d8\",\"Domainname\":\"\",\"User\":\"\",\"AttachStdin\":false,\"AttachStdout\":false,\"AttachStderr\":false,\"PortSpecs\":null,\"ExposedPorts\":null,\"Tty\":false,\"OpenStdin\":false,\"StdinOnce\":false,\"Env\":null,\"Cmd\":[\"/bin/sh\",\"-c\",\"#(nop) CMD [\\\"/hello\\\"]\"],\"Image\":\"a8219747be10611d65b7c693f48e7222c0bf54b5df8467d3f99003611afa1fd8\",\"Volumes\":null,\"VolumeDriver\":\"\",\"WorkingDir\":\"\",\"Entrypoint\":null,\"NetworkDisabled\":false,\"MacAddress\":\"\",\"OnBuild\":null,\"Labels\":{}},\"docker_version\":\"1.6.0\",\"config\":{\"Hostname\":\"d7c065bb52d8\",\"Domainname\":\"\",\"User\":\"\",\"AttachStdin\":false,\"AttachStdout\":false,\"AttachStderr\":false,\"PortSpecs\":null,\"ExposedPorts\":null,\"Tty\":false,\"OpenStdin\":false,\"StdinOnce\":false,\"Env\":null,\"Cmd\":[\"/hello\"],\"Image\":\"a8219747be10611d65b7c693f48e7222c0bf54b5df8467d3f99003611afa1fd8\",\"Volumes\":null,\"VolumeDriver\":\"\",\"WorkingDir\":\"\",\"Entrypoint\":null,\"NetworkDisabled\":false,\"MacAddress\":\"\",\"OnBuild\":null,\"Labels\":{}},\"architecture\":\"amd64\",\"os\":\"linux\",\"Size\":0}\n"
          },
          {
             "v1Compatibility": "{\"id\":\"a8219747be10611d65b7c693f48e7222c0bf54b5df8467d3f99003611afa1fd8\",\"created\":\"2015-04-17T22:01:38.329989482Z\",\"container\":\"d7c065bb52d847a4f27d39dfcce64186c56a6d112ad5732f5b7589faed064068\",\"container_config\":{\"Hostname\":\"d7c065bb52d8\",\"Domainname\":\"\",\"User\":\"\",\"AttachStdin\":false,\"AttachStdout\":false,\"AttachStderr\":false,\"PortSpecs\":null,\"ExposedPorts\":null,\"Tty\":false,\"OpenStdin\":false,\"StdinOnce\":false,\"Env\":null,\"Cmd\":[\"/bin/sh\",\"-c\",\"#(nop) ADD file:25f7e227e118ddc844c0dc0ec2dc765e8bb558fffb5bc546b4b609ad5a41c1fd in /\"],\"Image\":\"\",\"Volumes\":null,\"VolumeDriver\":\"\",\"WorkingDir\":\"\",\"Entrypoint\":null,\"NetworkDisabled\":false,\"MacAddress\":\"\",\"OnBuild\":null,\"Labels\":null},\"docker_version\":\"1.6.0\",\"config\":{\"Hostname\":\"d7c065bb52d8\",\"Domainname\":\"\",\"User\":\"\",\"AttachStdin\":false,\"AttachStdout\":false,\"AttachStderr\":false,\"PortSpecs\":null,\"ExposedPorts\":null,\"Tty\":false,\"OpenStdin\":false,\"StdinOnce\":false,\"Env\":null,\"Cmd\":null,\"Image\":\"\",\"Volumes\":null,\"VolumeDriver\":\"\",\"WorkingDir\":\"\",\"Entrypoint\":null,\"NetworkDisabled\":false,\"MacAddress\":\"\",\"OnBuild\":null,\"Labels\":null},\"architecture\":\"amd64\",\"os\":\"linux\",\"Size\":910}\n"
           }
       ],
       "schemaVersion": 1
    }

The topmost layer is the one with:

    "Cmd": [
        "/bin/sh",
        "-c",
        "#(nop) CMD [\"/hello\"]"
    ],

I think that because this line of the dockerfile doesn't change the filesystem it results in a duplicate layer reference.

Would the correct behavior here be: "for every entry in the history, make sure parent matches an ID of a subsequent entry"?